assets/src/dashboard/components/layout/provider.js

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/**
+ * External dependencies
+ */
+import React, {
+  createContext,
+  useMemo,
+  useRef,
+  useLayoutEffect,
+  useCallback,
+  useState,
+} from 'react';
+import PropTypes from 'prop-types';
+
+export const SQUISH_LENGTH = 90;
+export const SQUISH_CSS_VAR = '--squish-progress';
+
+const generateThreshold = (steps) => {
+  return Array.from({ length: steps + 1 }, (_, i) => i / steps);
+};
+
+export const LayoutContext = createContext(null);
+
+const Provider = ({ children }) => {
+  const [squishContentHeight, setSquishContentHeight] = useState(0);
+  const scrollFrameRef = useRef(null);
+  const squishDriverRef = useRef(null);
+
+  useLayoutEffect(() => {
+    const squishDriverEl = squishDriverRef.current;
+    const scrollFrameEl = scrollFrameRef.current;
+    if (!(squishDriverEl || scrollFrameEl)) {
+      return () => {};
+    }
+
+    const observer = new IntersectionObserver(
+      (entries) => {
+        entries.forEach((entry) => {
+          if (entry.intersectionRatio) {
+            const event = new CustomEvent('squish');
```

Comment:
Let's put a comment to polyfill this if we need to support IE. The constructor for CustomEvent does not exist in IE.